### PR TITLE
feat: naive client-side trial logs download [DET-3521]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

This changes buffering of trial logs from backend to client (the browser) and should eliminate the server crashes we've seen when downloading large trial logs. Same trial logs could still crash the browser but not the server. (the browser has no official API for streaming to file system more info on the ticket)
We can probably achieve a better memory footprint maybe using something like http://asmjs.org/ or reading more into how JS engines internally allocate and handle memory for different calls but we should see if we want to invest more time here or not.

## Test Plan

15643ms to buffer and download 3087007744 chars. total: ~3087007.744kB

client crashed at an attempt to buffer a 5GB Blob.

tested on chrome and firefox linux

## Commentary (optional)

we could look more into this and profile to see where we can get more wins.
dev usage of this depends on https://github.com/determined-ai/determined/pull/913

note: you'd have a problem downloading logs while the trial is running if you're using the react dev proxy and server. https://determinedai.atlassian.net/browse/DET-3638 will address that

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: webui/react/src/utils/browser.ts [hunk 1]
```diff
@@ -16,3 +19,63 @@ export const getCookie = (name: string): string | null => {
...
+  // https://jsben.ch/OJ3vo
+  // characters are utf16 encoded. v8 has different internal representations.
+  const MAX_PART_SIZE = 128 * Math.pow(2, 20); // 128m * CHAR_SIZE
+  const parts: BlobPart[] = [];
+  let downloadStringBuffer = '';
+  return consumeStream<DetSwagger.V1TrialLogsResponse>(
+    DetSwagger.ExperimentsApiFetchParamCreator().determinedTrialLogs(trialId),
+    (ev) => {
+      if (downloadStringBuffer.length > MAX_PART_SIZE) {
+        parts.push(downloadStringBuffer);
```
Comment (hamidzr) [inline webui/react/src/utils/browser.ts:53]: we can directly write into a preallocated `parts[I]` position could save a copy and a maximum of MAX_PART_SIZE

File: webui/react/src/utils/browser.ts [hunk 2]
```diff
@@ -16,3 +19,63 @@ export const getCookie = (name: string): string | null => {
...
+        downloadStringBuffer = '';
+      }
+      downloadStringBuffer += ev.message;
+    },
+  ).then(() => downloadText(`trial_${trialId}_logs.txt`, parts));
```
Comment (hamidzr) [inline webui/react/src/utils/browser.ts:58]: we need experiment details here to include in the filename to keep parity with the old download

File: webui/react/src/utils/browser.ts [hunk 3]
```diff
@@ -16,3 +20,66 @@ export const getCookie = (name: string): string | null => {
...
+const generateLogStringBuffer = (count: number, avgLength: number): string => {
+  const msg = new Array(avgLength).fill('a').join('') + '\n';
+  let stringBuffer = '';
+  for (let i=0; i<count; i++) {
+    stringBuffer += (i + msg);
+  }
+  return stringBuffer;
+};
+
+// TEMP TODO: remove me
```
Comment (hamidzr) [inline webui/react/src/utils/browser.ts:73]: probably remove the test/benchmark code?

File: webui/react/src/utils/browser.ts [hunk 4]
```diff
@@ -16,3 +20,66 @@ export const getCookie = (name: string): string | null => {
...
+  const MAX_PART_SIZE = 128 * Math.pow(2, 20); // 128m * CHAR_SIZE
+  const chunkCount = Math.ceil(numCharacters / MAX_PART_SIZE);
+  const parts = new Array(chunkCount).fill(0)
+    .map(() => generateLogStringBuffer(Math.pow(2, 20), 128));
+  downloadText('testlogs.txt', parts);
+  console.log(`${Date.now() - start}ms to buffer and download
+  ${chunkCount * MAX_PART_SIZE} chars. total: ~${chunkCount * MAX_PART_SIZE/1000}kB`);
+};
+
+// simulateLogsDownload(1e9);
```
Comment (hamidzr) [inline webui/react/src/utils/browser.ts:85]: @hkang1 @armandmcqueen it'd be nice to run this on different machines other than mine (AMD X86_64 and Linux) and see how practical it is?

File: webui/react/src/components/LogViewer.tsx [hunk 5]
```diff
@@ -13,14 +13,13 @@ import useScroll, { defaultScrollInfo } from 'hooks/useScroll';
...
 import { ansiToHtml, copyToClipboard, toRem } from 'utils/dom';
-import { openBlank } from 'utils/routes';
 
 import css from './LogViewer.module.scss';
 
 interface Props {
   debugMode?: boolean;
   disableLevel?: boolean;
-  downloadUrl?: string;
+  downloadHandler?: () => Promise<void>;
```
Comment (hkang1) [inline webui/react/src/components/LogViewer.tsx:22]: **nit:** `onDownload`?

File: webui/react/src/utils/browser.ts (hunk 1 above)
Comment (hkang1) [reply webui/react/src/utils/browser.ts:54]: **issue:** if the `downloadStringBuffer` doesn't exceed `MAX_PART_SIZE` nothing gets downloaded. Suggestion of pushing the final bits of `downloadStringBuffer` to `parts` **after** `consumeStream` finishes before calling `downloadText`.

File: webui/react/src/utils/browser.ts (hunk 3 above)
Comment (hkang1) [reply webui/react/src/utils/browser.ts:73]: **thought:** How about updating the `LogViewer.stories.tsx` to have a large view/download example?

File: webui/react/src/utils/browser.ts (hunk 4 above)
Comment (hkang1) [reply webui/react/src/utils/browser.ts:85]: `simulateLogsDownload` ran great for me on OSX

File: webui/react/src/pages/TrialLogs.tsx [hunk 6]
```diff
@@ -87,11 +86,23 @@ const TrialLogs: React.FC = () => {
     if (logsRef.current) logsRef.current?.addLogs(pollingLogsResponse.data);
   }, [ logIdRange, pollingLogsResponse.data ]);
 
+  const downloadLogs = useCallback(() => {
+    return downloadTrialLogs(id).catch(e => {
```
Comment (hkang1) [inline webui/react/src/pages/TrialLogs.tsx:90]: **nit:**
```
  return downloadTrialLogs(id).catch(e => {
    handleError({
      ...
    });
  });
```

Comment (hamidzr) [reply webui/react/src/pages/TrialLogs.tsx:90]: sidenote, on serious problem the page probably doesn't get here and crashes.

File: webui/react/src/utils/browser.ts (hunk 3 above)
Comment (hamidzr) [reply webui/react/src/utils/browser.ts:72]: that's a good idea.

Comment (hamidzr) [reply webui/react/src/utils/browser.ts]: I added this with configurable log size

Comment (hkang1) [reply webui/react/src/utils/browser.ts]: beautiful

**Pull request merged**